### PR TITLE
Implement startup behavior with StartUpMode

### LINKED ISSUE #11747: Mode Select Cluster's StartUpMode has no effect
#### Problem
 * expected behavior: according to the appclusters.pdf (1.8.5.4) The StartUpMode attribute SHALL define the desired startup mode for a device when it is supplied with power and this mode SHALL be reflected in the CurrentMode attribute.
 * actual behavior: StartUpMode has no effect.

 #### Proposed Solution
Add implementation in `src/app/clusters/mode-select-server/mode-select-server.cpp`

### PR DESCRIPTION
**Pull request by minhlez**
#### Problem
Fixes #11747

#### Change overview
* Change CurrentMode attribute to StartUpMode if they are both non-volatile and have different value
* Implement validation logic when updating StartUpMode 

#### Testing
How was this tested? (at least one bullet point required)
- Updated current unit test ModeSelect cluster in TestModeSelectCluster.yaml. Please note, however, that the power-up behavior can't be tested locally at the moment because the SystemCommand reboot needs CI to work.

### CODE REVIEW

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 1]
```diff
@@ -28,12 +28,22 @@
...
 #include <lib/support/CodeUtils.h>
 
 using namespace std;
 using namespace chip;
 using namespace chip::app;
 using namespace chip::app::Clusters;
+using namespace chip::app::Clusters::ModeSelect;
+using namespace chip::Protocols::InteractionModel;
+
+#ifndef IGNORE_MODE_SELECT_CLUSTER_START_UP_MODE
```
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:42]: Who would define this and where?  Should this be a CHIP_CONFIG setting?

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 2]
```diff
@@ -28,12 +28,22 @@
...
+static bool areStartUpModeServerAttributesNonVolatile(EndpointId endpoint);
+#endif
+
+Protocols::InteractionModel::Status verifyModeValue(const EndpointId endpointId, const uint8_t newMode);
```
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:46]:
```suggestion
InteractionModel::Status verifyModeValue(const EndpointId endpointId, const uint8_t newMode);
```

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml [hunk 3]
```diff
@@ -44,7 +44,7 @@ limitations under the License.
     <attribute side="server" code="0x0000" define="CURRENT_MODE"    type="INT8U"                                            writable="false" optional="false" reportable="true">CurrentMode</attribute>
     <attribute side="server" code="0x0001" define="SUPPORTED_MODES" type="ARRAY" entryType="ModeOptionStruct" length="255"  writable="false" optional="false">SupportedModes</attribute>
     <attribute side="server" code="0x0002" define="ON_MODE"         type="INT8U"                                            writable="true"  optional="true" default="255">OnMode</attribute>
-    <attribute side="server" code="0x0003" define="START_UP_MODE"   type="INT8U"                                            writable="false" optional="false">StartUpMode</attribute>
+    <attribute side="server" code="0x0003" define="START_UP_MODE"   type="INT8U"                                            writable="true" optional="true">StartUpMode</attribute>
```
Comment (bzbarsky-apple) [inline src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml:47]: This should be nullable.

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 4]
```diff
@@ -103,7 +113,119 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+        EmberAfStatus status = Attributes::StartUpMode::Get(endpointId, &startUpMode);
+        if (status == EMBER_ZCL_STATUS_SUCCESS)
+        {
+            // Initialise currentMode to 0
+            uint8_t currentMode = 0;
+            status            = Attributes::CurrentMode::Get(endpointId, &currentMode);
+            if (status == EMBER_ZCL_STATUS_SUCCESS)
+            {
+                if (startUpMode != currentMode) {
+                    status = Attributes::CurrentMode::Set(endpointId, startUpMode);
```
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:144]: How does this interact with OnMode?

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 2 above)
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:46]: Should be static or in anonymous namespace, no?

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 5]
```diff
@@ -103,7 +113,119 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+    return result;
+}
+
+/**
+ * Checks the new mode against the endpoint's supported modes.
+ * @param endpointId    endpointId of the endpoint
+ * @param newMode       value of the new mode
+ * @return              Success status if the value is valid; InvalidValue otherwise.
+ */
+Protocols::InteractionModel::Status verifyModeValue(const EndpointId endpointId, const uint8_t newMode)
```
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:221]:
```suggestion
InteractionModel::Status verifyModeValue(const EndpointId endpointId, const uint8_t newMode)
```

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 6]
```diff
@@ -103,7 +113,119 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+{
+    const ModeSelect::Structs::ModeOptionStruct::Type * modeOptionPtr;
+    EmberAfStatus checkSupportedModeStatus =
+        ModeSelect::getSupportedModesManager()->getModeOptionByMode(endpointId, newMode, &modeOptionPtr);
+    if (EMBER_ZCL_STATUS_SUCCESS != checkSupportedModeStatus)
+    {
+        return ToInteractionModelStatus(checkSupportedModeStatus);
+    }
+    return Protocols::InteractionModel::Status::Success;
```
Comment (bzbarsky-apple) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:230]:
```suggestion
    return InteractionModel::Status::Success;
```

File: src/app/tests/suites/TestModeSelectCluster.yaml [hunk 7]
```diff
@@ -88,3 +91,23 @@ tests:
...
+
+    - label: "Change to Unsupported StartUp Mode"
+      command: "writeAttribute"
+      attribute: "StartUpMode"
+      arguments:
+        value: 2
+      response:
+        error: CONSTRAINT_ERROR
+
+    - label: "Change to Supported StartUp Mode"
```
Comment (bzbarsky-apple) [inline src/app/tests/suites/TestModeSelectCluster.yaml:103]: This is not testing the actual startup behavior.  Please add tests for that?

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
@minhlez Please fix the restyle issues?

File: src/darwin/Framework/CHIP/templates/templates.json [hunk 8]
```diff
@@ -91,6 +91,16 @@
             "path": "CHIPAttributeTLVValueDecoder-src.zapt",
             "name": "Decode TLV attribute values into Objc objects",
             "output": "src/darwin/Framework/CHIP/zap-generated/CHIPAttributeTLVValueDecoder.mm"
+        },
+        {
+            "path": "../../../../../examples/chip-tool-darwin/templates/CHIPTestClustersObjc.zapt",
```
Comment (andy31415) [inline src/darwin/Framework/CHIP/templates/templates.json:96]: examples/chip-tool-darwin/templates/templates.json seems to alraedy contain these. Why this change?

**Review by andy31415 — DISMISSED** [review]
Codegen updates sis unclear. ChipTestClustersObjC is already included in examples/chip-tool-darwin/templates/templates.json

File: src/darwin/Framework/CHIP/templates/templates.json (hunk 8 above)
Comment (bzbarsky-apple) [inline src/darwin/Framework/CHIP/templates/templates.json:96]: This part is wrong.  Update to tip, where zap_regen_all updates chip-tool-darwin correctly, please.

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 1 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp:42]: - I followed the current pattern in [level-control.cpp](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/level-control/level-control.cpp#L74) and [on-off-server.cpp](https://github.com/project-chip/connectedhomeip/blob/master/src/app/clusters/on-off-server/on-off-server.cpp#L213). I assumed there is a mechanism to manage these flags. Let me know this is not the way.

File: src/app/tests/suites/TestModeSelectCluster.yaml (hunk 7 above)
Comment (minhlez) [reply src/app/tests/suites/TestModeSelectCluster.yaml:100]: This would require using the System's Reboot command, which was pointed out does not run locally with chip-tool in this [slack conversation](https://csamembers.slack.com/archives/G014G30SVV0/p1646854997668579). 
If this is for CI testing, I can include it in another PR, but it raises the question of how should I test it locally?

**Comment by minhlez** [comment]
> @minhlez Please fix the restyle issues?

Resolved in new push

File: src/darwin/Framework/CHIP/templates/templates.json (hunk 8 above)
Comment (minhlez) [reply src/darwin/Framework/CHIP/templates/templates.json]: - Was added as I thought it was needed, as discussed in this [slack conversationl](https://csamembers.slack.com/archives/G014G30SVV0/p1646935222327169)
- Removed in new push

Comment (minhlez) [reply src/darwin/Framework/CHIP/templates/templates.json]: I thought this was needed to fix an issue with codegen that I ran into, but as discussed in this [slack conversation](https://csamembers.slack.com/archives/G014G30SVV0/p1646935222327169), this won't be needed anymore.

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 2 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Resolved in new push

File: src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml (hunk 3 above)
Comment (minhlez) [reply src/app/zap-templates/zcl/data-model/chip/mode-select-cluster.xml]: Resolved in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 4 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: OnMode will have priority over this, as per the spec
> If this attribute is not null, the CurrentMode attribute SHALL be set to the StartUpMode value, when the server is powered up, except in the case when the OnMode attribute overrides the StartUpMode attribute.

Will be implemented in another PR

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 2 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated to static in new push.

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 5 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Resolved in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 6 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Resolved in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/clusters/mode-select-server/mode-select-server.cpp:42]: There might have been in the original Silicon Labs SDK, but I don't think there is in ours....  @selissia @jmartinez-silabs thoughts?

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: > Will be implemented in another PR

Please make sure there's an issue filed tracking that.

File: src/app/tests/suites/TestModeSelectCluster.yaml (hunk 7 above)
Comment (bzbarsky-apple) [reply src/app/tests/suites/TestModeSelectCluster.yaml:100]: You test it locally by doing:
```
./scripts/run_in_build_env.sh "./scripts/tests/run_test_suite.py --chip-tool /path/to/chip-tool --target TestModeSelectCluster run --iterations 1 --all-clusters-app /path/to/all-clusters-app"
```
which is what CI runs.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]
Wait, why are the efr sdk changes here?  There shouldn't be, I would think....

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 9]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+                                                                                  EmberAfAttributeType attributeType, uint16_t size,
+                                                                                  uint8_t * value)
+{
+    const EndpointId endpointId = attributePath.mEndpointId;
+    InteractionModel::Status result;
+
+    switch (attributePath.mAttributeId)
+    {
+    case ModeSelect::Attributes::StartUpMode::Id:
+        result = verifyModeValue(endpointId, value[0]);
```
Comment (du48s03) [inline src/app/clusters/mode-select-server/mode-select-server.cpp:206]:
```suggestion
        result = verifyModeValue(endpointId, *value);
```

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 4 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: We plan to reference this same issue. Once both this PR and this https://github.com/project-chip/connectedhomeip/pull/15888 have been merged.

File: src/app/tests/suites/TestModeSelectCluster.yaml (hunk 7 above)
Comment (minhlez) [reply src/app/tests/suites/TestModeSelectCluster.yaml:103]: This is great. I've included the test in the new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 10]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+    {
+        // Read the StartUpMode attribute and set the CurrentMode attribute
+        // The StartUpMode attribute SHALL define the desired startup behavior of a
+        // device when it is supplied with power and this state SHALL be
+        // reflected in the CurrentMode attribute.  The values of the StartUpMode
+        // attribute are listed below.
+
+        DataModel::Nullable<uint8_t> startUpMode;
+        EmberAfStatus status = Attributes::StartUpMode::Get(endpointId, startUpMode);
+        if (status == EMBER_ZCL_STATUS_SUCCESS && !startUpMode.IsNull() && startUpMode.HasValidValue())
```
[the remark this answers is not in the document]
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 11]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+                                       startUpMode.Value());
+                    }
+                }
+            }
+        }
+    }
+    else
+    {
+        emberAfPrintln(EMBER_AF_PRINT_DEBUG,
+                       "ModeSelect: Skipped initializing CurrentMode by StartUpMode because one of them is non-volatile");
```
[the remark this answers is not in the document]
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 12]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+        EmberAfStatus status = Attributes::StartUpMode::Get(endpointId, startUpMode);
+        if (status == EMBER_ZCL_STATUS_SUCCESS && !startUpMode.IsNull() && startUpMode.HasValidValue())
+        {
+            // Initialise currentMode to 0
+            uint8_t currentMode = 0;
+            status              = Attributes::CurrentMode::Get(endpointId, &currentMode);
+            if (status == EMBER_ZCL_STATUS_SUCCESS)
+            {
+                if (startUpMode.Value() != currentMode)
+                {
```
[the remark this answers is not in the document]
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 13]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+#endif
+}
+
+#ifndef IGNORE_MODE_SELECT_CLUSTER_START_UP_MODE
+/**
+ * Checks if StartUpMode and CurrentMode are non-volatile.
+ * @param endpointId    id of the endpoint to check
+ * @return true if both attributes are non-volatile; false otherwise.
+ */
+static bool areStartUpModeServerAttributesNonVolatile(EndpointId endpointId)
```
[the remark this answers is not in the document]
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp [hunk 14]
```diff
@@ -103,7 +113,120 @@ bool emberAfModeSelectClusterChangeToModeCallback(CommandHandler * commandHandle
...
+{
+    if (emberAfIsNonVolatileAttribute(endpointId, ModeSelect::Id, Attributes::CurrentMode::Id, true))
+    {
+        return emberAfIsNonVolatileAttribute(endpointId, ModeSelect::Id, Attributes::StartUpMode::Id, true);
+    }
+
+    return false;
```
[the remark this answers is not in the document]
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 9 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Updated in new push

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 4 above)
Comment (bzbarsky-apple) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: The point is, please make sure the issue is filed so it does not fall through the cracks.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by chrisdecenzo — APPROVED** [review]

**Review by du48s03 — APPROVED** [review]

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 1 above)
Comment (selissia) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: This is a leftover from Silabs cluster code where, yes, there was a mechanism to generate these flags based on user selections in a cluster configuration GUI. This is not applicable in Matter and we should not be using these flags here or in level-control.cpp and on-off-server.cpp

File: src/app/clusters/mode-select-server/mode-select-server.cpp (hunk 4 above)
Comment (minhlez) [reply src/app/clusters/mode-select-server/mode-select-server.cpp]: Created new issue to track this: https://github.com/project-chip/connectedhomeip/issues/16469

**Review by jmeg-sfy — APPROVED** [review]
